packages/client/src/client.ts

```diff
@@ -102,21 +107,141 @@ export class DefaultRoutingV1HttpApiClient implements RoutingV1HttpApiClient {
+    log('getIPNS starts: %c', pid)
+
+    const signal = anySignal([this.shutDownController.signal, options.signal, AbortSignal.timeout(this.timeout)])
+    const onStart = defer()
+    const onFinish = defer()
 
-    const pi = {
-      id: peer,
-      multiaddrs: ma,
-      protocols: []
+    void this.httpQueue.add(async () => {
+      onStart.resolve()
+      return onFinish.promise
+    })
+
+    try {
+      await onStart.promise
+
+      // https://specs.ipfs.tech/routing/http-routing-v1/
+      const resource = `${this.clientUrl}routing/v1/ipns/${pid.toCID().toString()}`
+      const getOptions = { headers: { Accept: 'application/vnd.ipfs.ipns-record' }, signal }
+      const a = await fetch(resource, getOptions)
+
+      if (a.body == null) {
+        throw new CodeError('GET ipns response had no body', 'ERR_BAD_RESPONSE')
+      }
+
+      const body = await toBuffer(toIt(a.body))
+      return unmarshal(body)
+    } finally {
+      signal.clear()
+      onFinish.resolve()
+      log('getIPNS finished: %c', pid)
+    }
+  }
+
+  async putIPNS (pid: PeerId, record: IPNSRecord, options: AbortOptions | undefined = {}): Promise<void> {
+    log('getIPNS starts: %c', pid)
+
+    const signal = anySignal([this.shutDownController.signal, options.signal, AbortSignal.timeout(this.timeout)])
+    const onStart = defer()
+    const onFinish = defer()
+
+    void this.httpQueue.add(async () => {
+      onStart.resolve()
+      return onFinish.promise
+    })
+
+    try {
+      await onStart.promise
+
+      const body = marshal(record)
+
+      // https://specs.ipfs.tech/routing/http-routing-v1/
+      const resource = `${this.clientUrl}routing/v1/ipns/${pid.toCID().toString()}`
+      const getOptions = { method: 'PUT', headers: { 'Content-Type': 'application/vnd.ipfs.ipns-record' }, body, signal }
+      const res = await fetch(resource, getOptions)
+      if (res.status !== 200) {
+        throw new CodeError('PUT ipns response had status other than 200', 'ERR_BAD_RESPONSE')
+      }
+    } finally {
+      signal.clear()
+      onFinish.resolve()
+      log('getIPNS finished: %c', pid)
+    }
+  }
+
+  #handleRecord (record: any): Record | null {
+    if (record.Schema === 'peer') {
+      // Peer schema can have additional, user-defined, fields.
+      record.ID = peerIdFromString(record.ID)
+      record.Addrs = record.Addrs.map(multiaddr)
+      return record
+    } else if (record.Schema === 'bitswap') {
+      // Bitswap schema cannot have additional fields.
+      return {
+        Schema: record.Schema,
+        Protocol: record.Protocol,
+        ID: peerIdFromString(record.ID),
+        Addrs: record.Addrs.map(multiaddr)
+      }
+    } else if (record.Schema !== '') {
+      // TODO: in Go, we send unknown schemas as an UnknownRecord. I feel like
+      // doing this here will make it harder. Is there a way in TypeScript
+      // to do something like if schema === 'bitswap' then it is a BitswapRecord?
```

Comment:
Question for maybe @ipfs/helia-dev

Reply:
Yes, see other comments about using `Schema` field as a disambiguator when `Record` is a union type.